### PR TITLE
nixos/sway: disable enableRealtime by default and remove nonworking patch

### PR DESCRIPTION
**Pull request by Artturin**
> FWIW my sway desktop started leaking capabilities:
> ```console
> $ getpcaps 0
> 0: cap_sys_nice=eip cap_setpcap+i
> ```
> Does the patch still work on your end?

> It doesn't work correctly and breaks pipewire-jack, as it relies on LD_LIBRARY_PATH and with this it is unset, previous PR also broke steam, but now it works for some reason, perhaps something was changed in the bubblewrap usage.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/programs/wayland/sway.nix [hunk 1]
```diff
@@ -45,7 +45,9 @@ in {
     enableRealtime = mkEnableOption (lib.mdDoc ''
       add CAP_SYS_NICE capability on `sway` binary for realtime scheduling
       privileges. This may improve latency and reduce stuttering, specially in
-      high load scenarios'') // { default = true; };
+      high load scenarios.
+      May not work.
```
Comment (thiagokokada) [inline nixos/modules/programs/wayland/sway.nix:49]: What does it mean it may not work? It works, just with caveats.

**Review by ghost — APPROVED** [review]
:+1: for removing NixOS-specific patches.  I run sway on non-NixOS and am mildly annoyed this was here to begin with...

File: nixos/modules/programs/wayland/sway.nix [hunk 2]
```diff
@@ -45,7 +45,9 @@ in {
     enableRealtime = mkEnableOption (lib.mdDoc ''
       add CAP_SYS_NICE capability on `sway` binary for realtime scheduling
       privileges. This may improve latency and reduce stuttering, specially in
-      high load scenarios'') // { default = true; };
+      high load scenarios.
+      Will break programs which depend on glibc `UNSECURE_ENVVARS'
+      <https://codebrowser.dev/glibc/glibc/sysdeps/generic/unsecvars.h.html>'');
```
Comment (Artturin) [inline nixos/modules/programs/wayland/sway.nix:50]: Made the comment better

File: nixos/modules/programs/wayland/sway.nix [hunk 3]
```diff
@@ -45,7 +45,9 @@ in {
     enableRealtime = mkEnableOption (lib.mdDoc ''
       add CAP_SYS_NICE capability on `sway` binary for realtime scheduling
       privileges. This may improve latency and reduce stuttering, specially in
-      high load scenarios'') // { default = true; };
+      high load scenarios.
+      Will break programs which depend on glibc's `UNSECURE_ENVVARS` set before launching sway
+      <https://codebrowser.dev/glibc/glibc/sysdeps/generic/unsecvars.h.html>'');
```
Comment (alyssais) [inline nixos/modules/programs/wayland/sway.nix:50]: Should we maybe also explain that it leaks capabilities?

**Review by kanashimia — APPROVED** [review]
Good. You can also add to the description something like "it currently relies on the security wrappers infrastructure, which [list caveats here], you could instead configure `rtprio` using `security.pam.loginLimits` without having to worry about those caveats" maybe with an example.
Well, I don't care much, looks good.

**Review by thiagokokada — APPROVED** [review]

**Comment by thiagokokada** [comment]
> :+1: for removing NixOS-specific patches.  I run sway on non-NixOS and am mildly annoyed this was here to begin with...

This patch was only applied if you used the NixOS module. In non-NixOS systems this patch was not applied.

**Comment by thiagokokada** [comment]
> Good. You can also add to the description something like "it currently relies on the security wrappers infrastructure, which [list caveats here], you could instead configure `rtprio` using `security.pam.loginLimits` without having to worry about those caveats" maybe with an example.
> Well, I don't care much, looks good.

We could simply reimplement the option using rtprio if so. Not sure why to overly complicate the description if we want to recommend another solution.

**Comment by thiagokokada** [comment]
> Good. You can also add to the description something like "it currently relies on the security wrappers infrastructure, which [list caveats here], you could instead configure `rtprio` using `security.pam.loginLimits` without having to worry about those caveats" maybe with an example. Well, I don't care much, looks good.

Ok, created a PR that reimplements this option in terms of `security.pam.loginLimits`, because it seems this is better solution anyway: breaks no programs and it is also compatible with Home-Manager. Add in the description the caveats.

I am fine merging any of those 2 PRs, but I think if we want to just document every possible breakage out there, it is better to just put the solution that is less painful for users. WDYT @Artturin?

**Review by eclairevoyant — COMMENTED** [review]
If we really must have the RT prio as part of the sway module, the alternative approach (not relying on capabilities) is more predictable.

However, do we really need to have this as part of the sway module? Or is that discussion out of scope? My concern is no matter how this is implemented, this is leaking additional "permissions" (for lack of better term)  to arbitrary programs.

**Comment by alyssais** [comment]
You make a good point that the PAM approach is actually not specific to Sway at all, so probably shouldn't live in the Sway module.  Since it's one line of config, maybe it's not necessary to have an option at all, and instead it could just be documented somewhere (wiki?) (with clear explanations of the security implications, of course!).

**Comment by kanashimia** [comment]
> You make a good point that the PAM approach is actually not specific to Sway at all, so probably shouldn't live in the Sway module. Since it's one line of config, maybe it's not necessary to have an option at all, and instead it could just be documented somewhere (wiki?) (with clear explanations of the security implications, of course!).

Was my point in the https://github.com/NixOS/nixpkgs/pull/263206#issuecomment-1777894022

**Comment by thiagokokada** [comment]
> You make a good point that the PAM approach is actually not specific to Sway at all, so probably shouldn't live in the Sway module.  Since it's one line of config, maybe it's not necessary to have an option at all, and instead it could just be documented somewhere (wiki?) (with clear explanations of the security implications, of course!).

In my opinion the Wiki is a bad place to document those things because it is less discoverable. I think the approach proposed in my PR is reasonable considering that the description has clear information about the implications. And of course the implementation can always be improved later.

But well, we can also keep the current approach. I am not against it.

**Comment by alyssais** [comment]
> In my opinion the Wiki is a bad place to document those things because it is less discoverable.

It would also be fine in the Nixpkgs or NixOS manual.

**Comment by thiagokokada** [comment]
> > In my opinion the Wiki is a bad place to document those things because it is less discoverable.
> 
> It would also be fine in the Nixpkgs or NixOS manual.

Ok, closed my other PR.

I think we can add the `security.pam.loginLimits` alternative to the description of this option then (like someone said above). I don't think this is the best option, but at least it should be discoverable and clear.

**Comment by alyssais** [comment]
> I think we can add the `security.pam.loginLimits` alternative to the description of this option then (like someone said above). I don't think this is the best option, but at least it should be discoverable and clear.

I definitely don't think we should have the security-wrapper based version at all — the PAM approach is definitely better.  Either we introduce a non-Sway-specific option to do the PAM thing, or we just document the PAM thing.

File: nixos/modules/programs/wayland/sway.nix (hunk 3 above)
Comment (thiagokokada) [inline nixos/modules/programs/wayland/sway.nix:50]:
```suggestion
      high load scenarios.
      
      Will break programs which depend on glibc's `UNSECURE_ENVVARS` set before launching sway
      <https://codebrowser.dev/glibc/glibc/sysdeps/generic/unsecvars.h.html>, and
      it is known to leak capabilities to other programs run inside Sway.
      
      As an alternative, you can set 
      `security.pam.loginLimits = [{ domain = "@users"; item = "rtprio"; type = "-"; value = 1; }]`
      in your config. This has none of the downsides described above, however any
      program running under "users" group will have permission to request real-time
      capabilities'');
```

**Comment by Artturin** [comment]
Meh let's just revert

**Comment by Artturin** [comment]
Why doesn't sway use rtkit btw?

**Comment by kanashimia** [comment]
> Wouldn't sway using rtkit fix the issues?

They didn't like the idea of having dbus in sway, search for original PR for adding cap_sys_nice support in sway.

**Comment by Artturin** [comment]
> > Wouldn't sway using rtkit fix the issues?
> 
> They didn't like the idea of having dbus in sway, search for original PR for adding cap_sys_nice support in sway.

For the interested
https://github.com/swaywm/sway/pull/6994
https://github.com/swaywm/sway/pull/6992

**Pull request closed without merging**